clear search on escape key

FroggieFrog opened the pull request:
- search text is cleared on escape key
- use (some kind of) MVVM for search text

Unfortunately contains PR #50.

AgmasGold commented:
I'm not sure if this change is needed - it was binding directly to the settings (I just added the bit on close as it only updates when focus is lost from the text box).

The PresetsSearchText isn't localised (and isn't UI text either), so doesn't belong in the Localization.MainWidnow class..

FroggieFrog commented:
It is cleaner to bind to a ViewModel. So there can be some unit tests in the future.

You are right that Localization.MainWindow is not a good place. But for the time being this is kind of the ViewModel for the MainWindow.
A cleaner architecture would be too massive at this point. So I would say to "expand" this class step by step to a full blown ViewModel and later rename it. Hence I would implement all new "features" in a MVVM style and chnage existing behaviour as it passes by.
I think the ultimate goal is a cleaner architecture and maybe a more standard approach to localization (resx?).

AgmasGold commented:
Yeah, I can see that.

I think resx stuff is based on the CurrentCulture of the thread, so I'm not 100% sure how we'd get that to work with the language selection. A better approach to localization is definitely needed.

AgmasGold commented:
Just doing some tests - looks like some of the functionality added in a previous merge has been removed (I think you are calling Reset() in the wrong places). The state of the TreeView should be saved before searching, and should return to that state when the search box is emptied.

AgmasGold commented:
had some issues at the start (hence the "revert" message, which I can't undo), I've restored the previous behaviour with restoring state when the search box is emptied.

FroggieFrog commented:
Sorry. Maybe the reverted search logic is caused by a missing merge with master. I don't know.

I tested the code and it seems to work.

I also added following features:
- clear search text when changing the language
- added a tooltip, so the user knows about that feature

Added loacalizations:
eng
{ "SearchToolTip", "ESC to clear search text" },
ger
{ "SearchToolTip", "ESC um Suchtext zu leeren" },
pol
{ "SearchToolTip", "ESC aby wyczyścić tekst do przeszukiwania" },
rus
{ "SearchToolTip", "ESC чтобы очистить текст поиска" },

AgmasGold commented:
I've added the localizations to the spreadsheet - also, I've called UpdateLayout() after the escape key is pressed. Unfortunately, without this, the caret position doesn't update until the user types some text again.

I tried setting .Text to "", InvalidateVisual(), and neither worked to reset the caret position.

![image](https://user-images.githubusercontent.com/48861520/57392475-c97f6000-71b8-11e9-8fe7-12ed6923ad0c.png)


After adding `TextBoxSearchPresets.UpdateLayout();`:

![image](https://user-images.githubusercontent.com/48861520/57392508-ddc35d00-71b8-11e9-8021-9233411be38f.png)

FroggieFrog commented:
Strange behaviour. In a simple testproject I was not able to reproduce it. It has to be some weird conflict with other logic.
But I can confirm that this is fixed now.

Could also be added for 8.3. 🚢
